File: M3Tools/Types/CustomersBindingSource.vb
Imports System.ComponentModel

Namespace Types
	Public Class CustomersBindingSource
		Inherits Windows.Forms.BindingSource

		Private _customers As New CustomersCollection

		Shadows ReadOnly Property SupportsFiltering As Boolean = True

		Shadows Property DataSource As CustomersCollection
			Get
				Return _customers
			End Get
			Set(value As CustomersCollection)
				_customers = value
			End Set
		End Property

		Shadows ReadOnly Property List As IList(Of Customer)
			Get
				Return _customers.Items
			End Get
		End Property

		'Shadows ReadOnly Property DataSource As CustomersCollection
		'	Get
		'		If DesignMode Then
		'			Return New CustomersCollection From {
		'				New Customer()
		'			}
		'		End If

		'		Return _customers
		'	End Get
		'End Property

		'Sub New()
		'	'MyBase.New(New CustomersCollection(), String.Empty)
		'	MyBase.New()
		'	DataSource = _customers
		'	ResetBindings(False)
		'End Sub

		Shadows Property Filter As String
			Get
				Return MyBase.Filter
			End Get
			Set(value As String)
				'If Not String.IsNullOrEmpty(value) Then
				'	value = $"[Name] like '%{value}%' OR [Email] like '%{value}%'"
				'End If

				MyBase.Filter = value
				_customers.Filter = value
			End Set
		End Property
	End Class
End Namespace

File: M3Tools/Types/CustomersBindingSource.cs
﻿using System.Collections.Generic;

namespace SPPBC.M3Tools.Data
{
    public class CustomersBindingSource : BindingSource<Types.Customer>
    {
		private Types.CustomerCollection _customers;

		public CustomersBindingSource() 
		{
			_customers = new Types.CustomerCollection();
			DataSource = _customers;
		}

		/// <summary>
		/// List of customers in the binding source
		/// </summary>
        public new IList<Types.Customer> List
        {
            get
            {
                return DataSource.Items;
            }
        }

		/// <summary>
		/// The data source for the binding source to bind from
		/// </summary>
		public new Types.CustomerCollection DataSource
		{
			get
			{
				return (Types.CustomerCollection)base.DataSource;
			}
			private set
			{
				base.DataSource = value;
			}
		}

		public override string Filter
		{ 
			get => base.Filter;
			set => base.Filter = value;
		}
	}
}
